### PR TITLE
Added Swift mocks part to unit test guidelines

### PR DESCRIPTION
**Pull request by RuudPuts**

### CODE REVIEW

File: swift/unit_tests.md [hunk 1]
```diff
@@ -237,4 +237,90 @@ class BoolSpec: QuickSpec {
     }
 }
 
+```
+
+## Writing Swift mocks
+
+### Terminology
+
+When it comes to writing mocks in Swift theres a great many ways developers do this, but we can never agree upon the naming.
```
Comment (marcopax) [inline swift/unit_tests.md]: theres a great many ways developers do this -> developers can do that in great many ways ?

File: swift/unit_tests.md [hunk 2]
```diff
@@ -237,4 +237,90 @@ class BoolSpec: QuickSpec {
...
+
+Fakes are objects that have working implementations, but not same as production one. Usually they take some shortcut and have simplified version of production code.
+![](https://cdn-images-1.medium.com/max/1600/0*snrzYwepyaPu3uC9.png)
+
+### Example mock from protocol
+
+As you can read, neither `...Stub` nor `...Mock` would cover the full load of a object we want to write. Our test-objects should be able to mimic a protocol by both setting stubs and registering received method calls. We don't want multiple files for a single Protocol in order to both mock & stub them, thus we want one object which can both stub and mock a protocol.
+
+The `...Double` suffix is coined as wel for this goal, derived from body-double, but this term on it's own only adds confusion. In the end calling your 'minimal' implementation of a class a `Mock` feels most natural, where a mock is build up from two parts;
+ - Captures: Basically the Mocks as described above. They register the captured method calls
```
Comment (marcopax) [inline swift/unit_tests.md]: I opened a point for adding Callbacks, as it seems more readable in a test especially for people without a long experience with testing.

```
struct Callbacks {
   var methodOne = false
   var methodTwo = false
}
```

And setting them to true when method is called.

I know that settings the Captures already give us information about a method being called or not, this suggestion is to improve readability of a test, like this example:

```
expect(myMock.captures.methods).toNot(beNil())
expect(myMock.callbacks.methods).to(beTrue)
```

Comment (RuudPuts) [reply swift/unit_tests.md]: I'm opposed to adding things to make it 'easier' for people without a lot of testing experience, we should learn them the correct way at once.

Also I would not like to do both the `bool` and a struct for the parameters as that's storing the same state twice. The argument for beginners could be used here as well "Beginners could easily miss one of the two".

Comment (RuudPuts) [reply swift/unit_tests.md]: Just noticed my example was wrong in this prospect, updating that. Please have another look afterwards.

File: swift/unit_tests.md (hunk 1 above)
Comment (RuudPuts) [reply swift/unit_tests.md]: Changed the text a bit

File: swift/unit_tests.md [hunk 3]
```diff
@@ -237,4 +237,94 @@ class BoolSpec: QuickSpec {
...
+ - Stubs: Stubs as described above
+ 
+ Both captures and stubs should be stored in a struct which can be instantiated without pasing parameters.
```
Comment (Bajocode) [inline swift/unit_tests.md]: 'pasing' --> 'passing' ?

Comment (FBajo) [inline swift/unit_tests.md]: "pasing" --> "passing"

File: swift/unit_tests.md [hunk 4]
```diff
@@ -237,4 +237,94 @@ class BoolSpec: QuickSpec {
...
+
+The `...Double` suffix is coined as wel for this goal, derived from body-double, but this term on it's own only adds confusion. In the end calling your 'minimal' implementation of a class a `Mock` feels most natural, where a mock is build up from two parts;
+ - Captures: Basically the Mocks as described above. They register the captured method calls
+ - Stubs: Stubs as described above
+ 
+ Both captures and stubs should be stored in a struct which can be instantiated without pasing parameters. 
+ `Stubs` speaks for itself. It contains stubs for both variables & method calls. Stubs should either be defined as an optional or with a default value.
+ Within `Captures` nested structs should be defined for each method to capture, in these nested structs the paramters of a function call shoud be stored. The definition of each variable within `Captures` should be optional. This way by checking the `Captures` variable for nil you can see wether or not a function is called, plus you can check the individual paramters passed to the call.
```
Comment (FBajo) [inline swift/unit_tests.md]: "shoud be stored" -> "should be stored"

Comment (FBajo) [reply swift/unit_tests.md]: "paramters" -> "parameters"

File: swift/unit_tests.md [hunk 5]
```diff
@@ -237,4 +237,94 @@ class BoolSpec: QuickSpec {
...
 
-```
\ No newline at end of file
+```
+
+## Writing Swift mocks
+
+### Terminology
+
+When it comes to writing mocks in Swift theres many ways for developers do this, but we can never agree upon the naming.
```
[the remark this answers is not in the document]
Comment (RuudPuts) [reply swift/unit_tests.md]: Wow, I had to read this comment 10x before noticing the 'to'

[the remark this answers is not in the document]
Comment (RuudPuts) [reply swift/unit_tests.md]: Fixed

File: swift/unit_tests.md (hunk 3 above)
Comment (RuudPuts) [reply swift/unit_tests.md]: Thanks, fixed

File: swift/unit_tests.md (hunk 4 above)
Comment (RuudPuts) [reply swift/unit_tests.md]: Thanks, fixed

**Pull request merged**